src/sys/futex.rs

```diff
@@ -0,0 +1,363 @@
+///         sleep(TIMEOUT);
+///         assert_eq!(futex.wake(1),Ok(1));
+///         // Wait for child process to exit
+///         unsafe {
+///             assert_eq!(libc::waitpid(child.as_raw(), std::ptr::null_mut(), 0), child.as_raw());
+///         }
+///     },
+///     ForkResult::Child => {
+///         let now = Instant::now();
+///         assert_eq!(futex.wait(0, Some(2 * TIMEOUT)),Ok(()));
+///         assert!(now.elapsed() > TIMEOUT && now.elapsed() < TIMEOUT + DELTA);
+///     }
+/// }
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct Futex<const PRIVATE: bool = false>(UnsafeCell<u32>);
+impl<const PRIVATE: bool> Futex<PRIVATE> {
+    const MASK: i32 = if PRIVATE { libc::FUTEX_PRIVATE_FLAG } else { 0 };
+    /// Constructs new futex with a given `val`.
+    pub fn new(val: u32) -> Self {
+        Self(UnsafeCell::new(val))
+    }
+    /// If the value of the futex:
+    /// - `== val`, the thread sleeps waiting for a [`Futex::wake`] call, in this case this thread
+    ///   is considered a waiter on this futex.
+    /// - `!= val`, then `Err` with [`Errno::EAGAIN`] is immediately returned.
+    ///
+    /// If the timeout is:
+    /// - `Some(_)` it specifies a timeout for the wait.
+    /// - `None` it will block indefinitely.
+    ///
+    /// Wraps [`libc::FUTEX_WAIT`].
+    pub fn wait(&self, val: u32, timeout: Option<Duration>) -> Result<()> {
+        let timespec = timeout.map(timespec);
+        let timespec_ptr = unwrap_or_null(timespec.as_ref());
+
+        let res = unsafe {
+            syscall(
+                SYS_futex,
+                self.0.get(),
+                Self::MASK | libc::FUTEX_WAIT,
+                val,
+                timespec_ptr,
+            )
+        };
+        Errno::result(res).map(drop)
+    }
+    /// Wakes at most `val` waiters.
+    ///
+    /// - `val == 1` wakes a single waiter.
+    /// - `val == u32::MAX` wakes all waiters.
+    ///
+    /// No guarantee is provided about which waiters are awoken. A waiter with a higher scheduling
+    /// priority is not guaranteed to be awoken in preference to a waiter with a lower priority.
+    ///
+    /// Wraps [`libc::FUTEX_WAKE`].
+    pub fn wake(&self, val: u32) -> Result<u32> {
+        let res = unsafe {
+            syscall(SYS_futex, self.0.get(), Self::MASK | libc::FUTEX_WAKE, val)
+        };
+        Errno::result(res).map(|x| u32::try_from(x).unwrap())
+    }
+    /// Creates a file descriptor associated with the futex.
+    ///
+    /// When [`Futex::wake`] is performed on the futex this file indicates being readable with
+    /// `select`, `poll` and `epoll`.
+    ///
+    /// The file descriptor can be used to obtain asynchronous notifications: if val is nonzero,
+    /// then, when another process or thread executes a FUTEX_WAKE, the caller will receive the
+    /// signal number that was passed in val.
+    ///
+    /// **Because it was inherently racy, this is unsupported from Linux 2.6.26 onward.**
```

Comment:
If  this is racy and has been removed since 2008, why wrap it?

Reply:
It is a part of the public futex API. I have no strong opinion either way on keeping or removing it here.